library/core/src/ptr/non_null.rs

```diff
@@ -1046,6 +1046,8 @@ impl<T: ?Sized> NonNull<T> {
     /// [`ptr::drop_in_place`]: crate::ptr::drop_in_place()
     #[inline(always)]
     #[stable(feature = "non_null_convenience", since = "1.80.0")]
+    #[kani::requires(ub_checks::can_dereference(self.as_ptr() as *const()))] // Ensure self is aligned, initialized, and valid for read
+    #[kani::requires(ub_checks::can_write(self.as_ptr() as *mut()))] // Ensure self is valid for write
```

Comment:
Why do these get there pointer cast to a different type?  As in, why `self.as_ptr() as *mut()` over `self.as_ptr()`?

Reply:
> Why do these get there pointer cast to a different type? As in, why `self.as_ptr() as *mut()` over `self.as_ptr()`?

Thanks for the suggestion @Eh2406 ! But there will be some error when using `self.as_ptr()`:
```
Kani Rust Verifier 0.56.0 (standalone)
error[E0277]: the trait bound `<T as ptr::metadata::Pointee>::Metadata: kani::mem::PtrProperties<T>` is not satisfied
    --> /Users/danieltu/Developer/verify-rust-std/library/core/src/ptr/non_null.rs:1050:43
     |
1050 |     #[kani::requires(ub_checks::can_write(self.as_ptr()))] // Ensure self is valid for write
     |                      -------------------- ^^^^^^^^^^^^^ the trait `kani::mem::PtrProperties<T>` is not implemented for `<T as ptr::metadata::Pointee>::Metadata`
     |                      |
     |                      required by a bound introduced by this call
     |
note: required by a bound in `kani::mem::can_write`
    --> /Users/danieltu/Developer/verify-rust-std/library/core/src/lib.rs:423:1
     |
423  | kani_core::kani_lib!(core);
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^
     | |
     | required by a bound in this function
     | required by this bound in `can_write`
     = note: this error originates in the macro `kani_core::kani_mem` which comes from the expansion of the macro `kani_core::kani_lib` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider further restricting the associated type
     |
1051 |     pub unsafe fn drop_in_place(self) where <T as ptr::metadata::Pointee>::Metadata: kani::mem::PtrProperties<T> {
     |                                       ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

For more information about this error, try `rustc --explain E0277`.
error: Failed to execute cargo (exit status: 101). Found 1 compilation errors.
```

I think maybe that's because `can_write` takes `*mut T` as input?